### ISSUE TITLE
app.CRITICAL: An exception occurred. Unknown column scripts in tl_layout bei Migration Contao 3.5.28 auf 4.4.2

### ISSUE THREAD

**Issue by mpunkt** [contao/core-bundle#1006]
Migration einer bestehenden Contao 3.5.28 Installation auf 4.4.2.  Installtool bricht mit Fehler ab.

**Comment by fritzmg** [comment]
Have you confirmed that your database is up to date with Contao 3.5.28 before starting the migration?

**Comment by mpunkt** [comment]
Yes.



Von: Fritz Michael Gschwantner [mailto:notifications@github.com]
Gesendet: Sonntag, 6. August 2017 19:32
An: contao/core-bundle
Cc: mpunkt; Author
Betreff: Re: [contao/core-bundle] app.CRITICAL: An exception occurred. Unknown column scripts in tl_layout bei Migration Contao 3.5.28 auf 4.4.2 (#1006)



Have you confirmed that your database is up to date with Contao 3.5.28 before starting the migration?

**Comment by fritzmg** [comment]
`tl_layout.scripts` definitely exists in Contao 3.5.28 - so if it doesn't while migrating, something went wrong during the migration process. You should seek help on the community forums.

**Comment by mpunkt** [comment]
No! There is only a column script!



Von: Fritz Michael Gschwantner [mailto:notifications@github.com]
Gesendet: Sonntag, 6. August 2017 19:52
An: contao/core-bundle
Cc: mpunkt; Author
Betreff: Re: [contao/core-bundle] app.CRITICAL: An exception occurred. Unknown column scripts in tl_layout bei Migration Contao 3.5.28 auf 4.4.2 (#1006)



tl_layout.scripts definitely exists in Contao 3.5.28 - so if it doesn't while migrating, something went wrong during the migration process. You should seek help on the community forums.

**Comment by Aybee** [comment]
I noticed the same thinkg last week. there only ist tl_layout.**script**, not tl_layout.**scripts** in Contao 3.5.28.

**Comment by fritzmg** [comment]
That's normal. The upgrade script of the installation-bundle adds the `scripts` field to the `tl_layout` database.

**Comment by fritzmg** [comment]
@mpunkt to find out more about your issue, you should run the migration with `app_dev.php`.

**Comment by Aybee** [comment]
Yes, but the error occurs during the update and we can not go on with the update cause this field is missing. I e.g. copied script manually to scripts.

**Comment by fritzmg** [comment]
Where exactly does the error occur? Use `app_dev.php`

**Comment by mpunkt** [comment]
My migration steps where:
1. Check C3.5.28 database with install.php => status where Ok.
2. Install C 4.2.2 ( curl -L ... | tar x ... and then https://<url>/contao/install ). At this point the tl_layout has a column 'script" and(!) a column 'scripts'.
3. Import database from C3.5.28 installation and run https://<url>/contao/install. Now the error occurs.
4. alter table tl_layout add column scripts ... Then https://<url>/contao/install works.

**Comment by ausi** [comment]
@mpunkt I think you should swap your steps two and tree: Import the C3.5.28 database before running the C4.2.2 install tool.

**Comment by fritzmg** [comment]
Indeed, you must start with the database of Contao 3 and not install Contao 4 without it.

**Comment by mpunkt** [comment]
I did it. But there is the same error "An exception occurred while executing 'SELECT id, scripts FROM tl_layout': Column not found"

**Comment by Aybee** [comment]
I also did it that way. Installed C4 **without** generating the tables with installtool, as the DB (from C3.5) was already there. Now when it comes to the install tool part, (here the DB gets checked and neccessary changes should be shown) the error occured.

**Comment by fritzmg** [comment]
Post the full stack trace.

**Comment by ausi** [comment]
The `SELECT id, scripts FROM tl_layout` could come from [src/Database/Version440Update.php](https://github.com/contao/installation-bundle/blob/c0bdff7b8556c7966926d7f202e05fc7fc7e9127/src/Database/Version440Update.php#L44). Maybe the update scripts get executed in the wrong order, or the 4.4.0 update should check if it should use `scripts` or `script`.

**Comment by mpunkt** [comment]
How can i get the 'full stacke trace'?

**Comment by fritzmg** [comment]
> How can i get the 'full stacke trace'?

Open the Install Tool like so: `example.org/app_dev.php/contao/install`, then click on "Stack Trace".

**Comment by mpunkt** [comment]
I get "You are not allowed to access this file."

**Comment by fritzmg** [comment]
In order to use the `app_dev.php`, you must first configure it. On the console run the following command:
```
vendor/bin/contao-console contao:install-web-dir --user=USER --password=PASSWORD
```
Replace `USER` and `PASSWORD` with a desired username and password.

**Comment by mpunkt** [comment]
**Stack Trace [3/3] InvalidFieldNameException** 
Doctrine\DBAL\Exception\InvalidFieldNameException:
An exception occurred while executing 'SELECT id, scripts FROM tl_layout':

SQLSTATE[42S22]: Column not found: 1054 Unknown column 'scripts' in 'field list'

  at vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:71
  at Doctrine\DBAL\Driver\AbstractMySQLDriver->convertException('An exception occurred while executing \'SELECT id, scripts FROM tl_layout\':SQLSTATE[42S22]: Column not found: 1054 Unknown column \'scripts\' in \'field list\'', object(PDOException))
     (vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:128)
  at Doctrine\DBAL\DBALException::driverExceptionDuringQuery(object(Driver), object(PDOException), 'SELECT id, scripts FROM tl_layout')
     (vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:964)
  at Doctrine\DBAL\Connection->query('SELECT id, scripts FROM tl_layout')
     (vendor/contao/installation-bundle/src/Database/Version440Update.php:44)
  at Contao\InstallationBundle\Database\Version440Update->run()
     (vendor/contao/installation-bundle/src/Controller/InstallationController.php:347)
  at Contao\InstallationBundle\Controller\InstallationController->runDatabaseUpdates()
     (vendor/contao/installation-bundle/src/Controller/InstallationController.php:99)
  at Contao\InstallationBundle\Controller\InstallationController->installAction()
  at call_user_func_array(array(object(InstallationController), 'installAction'), array())
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:153)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:68)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:171)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/app_dev.php:62)

**Comment by fritzmg** [comment]
May be the Finder needs an additional `->sortByName()` to be sure?

**Comment by mpunkt** [comment]
I can't answer this question.

**Comment by ausi** [comment]
I was able to reproduce this issue by sorting the update files in reverse order. So it looks like this issue is caused by the `Finder` class that returns the files in a random order.

> @ausi May be the Finder needs an additional `->sortByName()` to be sure?

I agree, we should add this in [installation-bundle/src/Controller/InstallationController.php:336](https://github.com/contao/installation-bundle/blob/c0bdff7b8556c7966926d7f202e05fc7fc7e9127/src/Controller/InstallationController.php#L336), this should solve the issue.

**Comment by Aybee** [comment]
Thank you for your effort of trying to reproduce the error and find out the cause. 💯

**Comment by mpunkt** [comment]
It works. Thank you very much!!!

**Comment by leofeyer** [comment]
Fixed in contao/installation-bundle@abcac12bd49ab04f0ff766e3c3d0679277f65b54.
